File: DynamicProgramming/123_BestTimetoBuyandSellStockIII.py
class Solution(object):

    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        if n <= 1:
            return 0
        max_to_here = [0] * n
        max_from_here = [0] * n

        min_val = prices[0]
        for i in range(1, n - 1):
            max_to_here[i] = max(max_to_here[i - 1], prices[i] - min_val)
            min_val = min(prices[i], min_val)

        max_val = prices[n - 1]
        for i in range(n - 2, -1, -1):
            max_from_here[i] = max(max_from_here[i + 1], max_val - prices[i])
            max_val = max(prices[i], max_val)

        max_pro = max_from_here[0]
        for i in range(0, n - 1):
            max_pro = max(max_pro, max_to_here[i] + max_from_here[i + 1])

        return max_pro

File: DynamicProgramming/123_BestTimetoBuyandSellStockIII.java
public class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if (n <= 1) return 0;
        int[] maxToHere = new int[n];
        int[] maxFromHere = new int[n];
        int minVal = prices[0];
        int maxVal = prices[n - 1];
        for (int i=1; i<n; i++) {
            if (prices[i] <= minVal) {
                minVal = prices[i];
                maxToHere[i] = maxToHere[i - 1];
            } else {
                maxToHere[i] = Math.max(maxToHere[i - 1], prices[i] - minVal);
            }
        }
        
        for (int i = n-2; i>=0; i--) {
            if (prices[i] >= maxVal) {
                maxVal = prices[i];
                maxFromHere[i] = maxFromHere[i + 1];
            } else {
                maxFromHere[i] = Math.max(maxFromHere[i + 1], maxVal - prices[i]);
            }
        }
        int mprofit = maxFromHere[0];
        for (int i=0; i<n-1; i++) {
            mprofit = Math.max(mprofit, maxToHere[i] + maxFromHere[i+1]);
        }
        return mprofit;
    }
}